tests/smoke_tests/run_smoke_test.py

```diff
@@ -0,0 +1,182 @@
+        )
+
+    logger.info("All clients finished execution")
+
+    # Collecting the server output when its process finish
+    full_server_output = await _wait_for_process_to_finish_and_retrieve_logs(server_process, "Server")
+
+    logger.info("Server has finished execution")
+
+    # server assertions
+    assert "error" not in full_server_output.lower(), (
+        f"Full output:\n{full_server_output}\n" + "[ASSERT ERROR] Error message found for server."
+    )
+    assert f"evaluate_round {config['n_server_rounds']}" in full_server_output, (
+        f"Full output:\n{full_server_output}\n" + "[ASSERT ERROR] Last FL round message not found for server."
+    )
+    assert "FL finished" in full_server_output, (
+        f"Full output:\n{full_server_output}\n" + "[ASSERT ERROR] FL finished message not found for server."
+    )
+    assert all(
+        message in full_server_output
+        for message in [
+            "app_fit: losses_distributed",
+            "app_fit: metrics_distributed_fit",
+            "app_fit: metrics_distributed",
+            "app_fit: losses_centralized",
+            "app_fit: metrics_centralized",
+        ]
+    ), f"Full output:\n{full_server_output}\n[ASSERT ERROR] Metrics message not found for server."
+
+    # client assertions
+    for i in range(len(full_client_outputs)):
+        assert "error" not in full_client_outputs[i].lower(), (
+            f"Full client output:\n{full_client_outputs[i]}\n" + f"[ASSERT ERROR] Error message found for client {i}."
+        )
+        assert f"Current FL Round: {config['n_server_rounds']}" in full_client_outputs[i], (
+            f"Full client output:\n{full_client_outputs[i]}\n"
+            + f"[ASSERT ERROR] Last FL round message not found for client {i}."
+        )
+        assert "Disconnect and shut down" in full_client_outputs[i], (
+            f"Full client output:\n{full_client_outputs[i]}\n"
+            + f"[ASSERT ERROR] Shutdown message not found for client {i}."
+        )
+
+    logger.info("All checks passed. Test finished.")
+
+
+def _preload_dataset(dataset_path: str, config: Config) -> None:
+    if "mnist" in dataset_path:
+        logger.info("Preloading MNIST dataset...")
+
+        # Working around MNIST download issue
+        # https://github.com/pytorch/vision/issues/1938
+        opener = urllib.request.build_opener()
+        opener.addheaders = [("User-agent", "Mozilla/5.0")]
+        urllib.request.install_opener(opener)
+
+        # Creating a client and getting the data loaders will trigger
+        # the dataset's download
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        client = MnistFedProxClient(Path(dataset_path), [Accuracy()], device)
+        client.get_data_loaders(config)
+
+        logger.info("Finished preloading MNIST dataset")
+    else:
+        logger.info("Preload not supported for specified dataset. Skipping.")
```

Comment:
Should we just throw an error here rather than proceeding? I imagine we'd blow up somewhere else, so might as well do it here a bit earlier? 😂

Reply:
I like the thinking, but the MNIST error seems to be specific to the server hosting it (details in the issue I linked in the code comment above). It's possible we don't need to preload other datasets, I'll find out about that when I add the other tests.

Reply:
Ah I see. That makes sense.